contracts/finance/Vesting.sol

```diff
@@ -0,0 +1,284 @@
+        return _vestings[vestingId_];
+    }
+
+    function getVestings(address beneficiary_) public view virtual returns (VestingData[] memory) {
+        uint256[] memory ids_ = _beneficiaryIds[beneficiary_].values();
+        VestingData[] memory beneficiaryVestings_ = new VestingData[](ids_.length);
+
+        for (uint256 i = 0; i < ids_.length; i++) {
+            beneficiaryVestings_[i] = _vestings[ids_[i]];
+        }
+
+        return beneficiaryVestings_;
+    }
+
+    function getVestingIds(address beneficiary_) public view virtual returns (uint256[] memory) {
+        return _beneficiaryIds[beneficiary_].values();
+    }
+
+    function getWithdrawableAmount(uint256 vestingId_) public view virtual returns (uint256) {
+        VestingData storage _vesting = _vestings[vestingId_];
+        Schedule storage _schedule = _schedules[_vesting.scheduleId];
+
+        return _getWithdrawableAmount(_vesting, _schedule, block.timestamp);
+    }
+
+    function getVestedAmount(uint256 vestingId_) public view virtual returns (uint256) {
+        VestingData storage _vesting = _vestings[vestingId_];
+        Schedule storage _schedule = _schedules[_vesting.scheduleId];
+
+        return _getVestedAmount(_vesting, _schedule, block.timestamp);
+    }
+
+    function _getWithdrawableAmount(
+        VestingData storage vesting,
+        Schedule storage schedule,
+        uint256 timestamp_
+    ) internal view virtual returns (uint256) {
+        return
+            _vestingCalculation(
+                schedule,
+                vesting.vestingAmount,
+                vesting.vestingStartTime,
+                timestamp_
+            ) - vesting.paidAmount;
+    }
+
+    function _getVestedAmount(
+        VestingData storage vesting,
+        Schedule storage schedule,
+        uint256 timestamp_
+    ) internal view virtual returns (uint256) {
+        return
+            _vestingCalculation(
+                schedule,
+                vesting.vestingAmount,
+                vesting.vestingStartTime,
+                timestamp_
+            );
+    }
+
+    function _createVesting(VestingData memory vesting_) internal virtual returns (uint256) {
+        require(
+            vesting_.vestingStartTime > 0,
+            "VestingWallet: cannot create vesting for zero time"
+        );
+        require(
+            vesting_.vestingAmount > 0,
+            "VestingWallet: cannot create vesting for zero amount"
+        );
+        require(
+            vesting_.beneficiary != address(0),
+            "VestingWallet: cannot create vesting for zero address"
+        );
+
+        Schedule storage _schedule = _schedules[vesting_.scheduleId];
+
+        require(
+            vesting_.vestingStartTime +
+                _schedule.scheduleData.durationInPeriods *
+                _schedule.scheduleData.secondsInPeriod >
+                block.timestamp,
+            "VestingWallet: cannot create vesting for past date"
+        );
+
+        _vestingId++;
```

Comment:
_vestingId storage variable usage can probably be optimized

Reply:
![image](https://github.com/dl-solarity/solidity-lib/assets/94608729/2a334f49-6170-4231-9d4c-b6589428a5bc)
like this?

Reply:
Nope. You are reading it like 5 times in a row. Copy it to stack then assign to storage